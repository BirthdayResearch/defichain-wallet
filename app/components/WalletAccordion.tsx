import React, { useState } from 'react'
import { ThemedIcon, ThemedScrollView, ThemedText, ThemedView, ThemedSectionTitle } from './themed'
import Accordion from 'react-native-collapsible/Accordion'
import { tailwind } from '@tailwind'
import { useThemeContext } from '@contexts/ThemeProvider'

interface AccordionProps {
  testID?: string
  title: string
  content: AccordionContent[]
}

export interface AccordionContent {
  title: string
  content: string
}

interface ActiveSessions {
  activeSessions: number[] | string[]
}

export function WalletAccordion (props: AccordionProps): JSX.Element {
  const { isLight } = useThemeContext()
  const initialContent: ActiveSessions = {
    activeSessions: []
  }
  const [activeContent, setActiveContent] = useState(initialContent)
  const isLastContent = (index: number): boolean => {
    return index === props.content.length - 1
  }

  return (
    <ThemedScrollView
      testID={props.testID}
    >
      <ThemedSectionTitle
        style={tailwind('mt-8 text-xs font-medium')}
        light={tailwind('text-gray-400')}
        dark={tailwind('text-gray-500')}
        text={props.title}
      />

      <Accordion
        containerStyle={[tailwind('border rounded-lg mt-2 overflow-hidden'), isLight ? tailwind('bg-white border-gray-200') : tailwind('bg-black border-gray-700')]}
        underlayColor='transparent'
        sections={props.content}
        renderHeader={(prop, index, isActive) => {
          return (
            <ThemedView
              light={tailwind('border-gray-200')}
              dark={tailwind('border-gray-700')}
              style={[tailwind('p-4 flex-row items-center justify-between'), !isActive && !isLastContent(index) && tailwind('border-b'), isActive && tailwind('pb-1')]}
            >
              <ThemedText
                style={[
                  tailwind('text-sm flex-1'),
                  isActive && tailwind('font-semibold')
                ]}
                light={isActive ? tailwind('text-gray-900') : tailwind('text-gray-500')}
                dark={isActive ? tailwind('text-gray-50') : tailwind('text-gray-400')}
              >
                {prop.title}
              </ThemedText>
              <ThemedIcon
                iconType='MaterialIcons'
                name={isActive ? 'arrow-drop-up' : 'arrow-drop-down'}
                size={24}
                light={isActive ? tailwind('text-gray-600') : tailwind('text-gray-400')}
                dark={isActive ? tailwind('text-gray-300') : tailwind('text-gray-500')}
              />
            </ThemedView>

          )
        }}
        renderContent={(prop, index, isActive) => {
          return (
            <ThemedView
              style={[tailwind('p-4 pt-0'), isActive && !isLastContent(index) && tailwind('border-b')]}
              light={tailwind('border-gray-200')}
              dark={tailwind('border-gray-700')}
            >
              <ThemedText
                style={tailwind('text-sm')}
                light={tailwind('text-gray-600')}
                dark={tailwind('text-gray-300')}
              >
                {prop.content}
              </ThemedText>
            </ThemedView>
          )
        }}
        onChange={(activeSessions) => {
          setActiveContent({ activeSessions })
        }}
        activeSections={activeContent.activeSessions}
      />
    </ThemedScrollView>
  )
}
