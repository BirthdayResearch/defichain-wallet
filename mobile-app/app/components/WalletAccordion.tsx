import { useState } from 'react'
import { ThemedIcon, ThemedScrollView, ThemedSectionTitle, ThemedText, ThemedView } from './themed'
import Accordion from 'react-native-collapsible/Accordion'
import { tailwind } from '@tailwind'
import { View } from 'react-native'
import { useThemeContext } from '@shared-contexts/ThemeProvider'

interface AccordionProps {
  testID?: string
  title: string
  content: AccordionContent[]
  activeSections?: number[] | string[]
}

export interface AccordionContent {
  title: string
  content: Array<{
    text: string
    type: 'bullet' | 'paragraph'
  }>
}

export function WalletAccordion (props: AccordionProps): JSX.Element {
  const { isLight } = useThemeContext()
  const [activeSections, setActiveSections] = useState<number[] | string[]>(props.activeSections ?? [])
  const isLastContent = (index: number): boolean => {
    return index === props.content.length - 1
  }

  return (
    <ThemedScrollView
      testID={props.testID}
    >
      <ThemedSectionTitle
        style={tailwind('mt-8 text-xs font-medium')}
        light={tailwind('text-gray-400')}
        dark={tailwind('text-gray-500')}
        text={props.title}
      />

      <Accordion
        containerStyle={[tailwind('border rounded-lg mt-2 overflow-hidden'), isLight ? tailwind('bg-white border-gray-200') : tailwind('bg-black border-gray-700')]}
        underlayColor='transparent'
        sections={props.content}
        renderHeader={(prop, index, isActive) => {
          return (
            <ThemedView
              light={tailwind('border-gray-200')}
              dark={tailwind('border-gray-700')}
              style={[tailwind('p-4 flex-row items-center justify-between'), !isActive && !isLastContent(index) && tailwind('border-b'), isActive && tailwind('pb-1')]}
            >
              <ThemedText
                style={[
                  tailwind('text-sm flex-1'),
                  isActive && tailwind('font-semibold')
                ]}
                light={isActive ? tailwind('text-gray-900') : tailwind('text-gray-500')}
                dark={isActive ? tailwind('text-gray-50') : tailwind('text-gray-400')}
              >
                {prop.title}
              </ThemedText>
              <ThemedIcon
                iconType='MaterialIcons'
                name={isActive ? 'arrow-drop-up' : 'arrow-drop-down'}
                size={24}
                light={isActive ? tailwind('text-gray-600') : tailwind('text-gray-400')}
                dark={isActive ? tailwind('text-gray-300') : tailwind('text-gray-500')}
              />
            </ThemedView>

          )
        }}
        renderContent={(prop, index, isActive) => {
          return (
            <ThemedView
              style={[tailwind('p-4 pt-0'), isActive && !isLastContent(index) && tailwind('border-b')]}
              light={tailwind('border-gray-200')}
              dark={tailwind('border-gray-700')}
            >
              {prop.content.map(({ text, type }) => (
                <View key={text} style={tailwind('flex-row justify-start')}>
                  {type === 'bullet' && (
                    <ThemedText style={tailwind('w-1/12 font-bold text-sm')}>{'\u2022'}</ThemedText>
                  )}
                  <ThemedText
                    key={text}
                    style={tailwind('flex-1 text-sm')}
                    light={tailwind('text-gray-600')}
                    dark={tailwind('text-gray-300')}
                  >
                    {text}
                  </ThemedText>
                </View>
              ))}
            </ThemedView>
          )
        }}
        onChange={(activeSections) => {
          setActiveSections(activeSections)
        }}
        activeSections={activeSections}
      />
    </ThemedScrollView>
  )
}
