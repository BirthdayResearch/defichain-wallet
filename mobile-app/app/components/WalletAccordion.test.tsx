import { render } from '@testing-library/react-native'

import { WalletAccordion } from './WalletAccordion'

jest.mock('@shared-contexts/ThemeProvider')

describe('wallet accordion', () => {
  it('should render', async () => {
    const rendered = render(
      <WalletAccordion
        content={[
          {
            title: 'foo',
            content: [{
              text: 'foo',
              type: 'paragraph'
            }, {
              text: 'bar',
              type: 'bullet'
            }]
          }
        ]}
        title='foo'
      />
    )
    expect(rendered.toJSON()).toMatchSnapshot()
  })
})
