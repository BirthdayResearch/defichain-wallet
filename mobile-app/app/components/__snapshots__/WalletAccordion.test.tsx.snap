// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wallet accordion should render 1`] = `
<RCTScrollView
  style={
    Array [
      undefined,
      Object {
        "backgroundColor": "rgba(245, 245, 245, 1)",
      },
    ]
  }
>
  <View>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "RegularFont",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
          Array [
            Object {
              "fontFamily": "MediumFont",
              "fontSize": 12,
              "fontWeight": "500",
              "lineHeight": 16,
              "marginTop": 32,
            },
            Object {
              "color": "rgba(163, 163, 163, 1)",
            },
          ],
        ]
      }
      text="foo"
    >
      foo
    </Text>
    <View
      style={
        Array [
          Object {
            "borderBottomLeftRadius": 8,
            "borderBottomRightRadius": 8,
            "borderBottomWidth": 1,
            "borderLeftWidth": 1,
            "borderRightWidth": 1,
            "borderTopLeftRadius": 8,
            "borderTopRightRadius": 8,
            "borderTopWidth": 1,
            "marginTop": 8,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "rgba(255, 255, 255, 1)",
            "borderBottomColor": "rgba(229, 229, 229, 1)",
            "borderLeftColor": "rgba(229, 229, 229, 1)",
            "borderRightColor": "rgba(229, 229, 229, 1)",
            "borderTopColor": "rgba(229, 229, 229, 1)",
          },
        ]
      }
    >
      <View
        style={Object {}}
      >
        <View
          accessibilityState={
            Object {
              "expanded": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            style={
              Array [
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "paddingBottom": 16,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "paddingTop": 16,
                  },
                  false,
                  false,
                ],
                Object {
                  "borderBottomColor": "rgba(229, 229, 229, 1)",
                  "borderLeftColor": "rgba(229, 229, 229, 1)",
                  "borderRightColor": "rgba(229, 229, 229, 1)",
                  "borderTopColor": "rgba(229, 229, 229, 1)",
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "RegularFont",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "lineHeight": 24,
                  },
                  Array [
                    Array [
                      Object {
                        "flexBasis": "0%",
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "fontSize": 14,
                        "lineHeight": 20,
                      },
                      false,
                    ],
                    Object {
                      "color": "rgba(115, 115, 115, 1)",
                    },
                  ],
                ]
              }
            >
              foo
            </Text>
            <Text />
          </View>
        </View>
        <View
          pointerEvents="none"
          style={
            Object {
              "height": 0,
              "overflow": "hidden",
            }
          }
        >
          <View
            forwardedRef={[Function]}
            onLayout={[Function]}
            style={Object {}}
          >
            <View
              style={
                Array [
                  Array [
                    Object {
                      "paddingBottom": 16,
                      "paddingLeft": 16,
                      "paddingRight": 16,
                      "paddingTop": 0,
                    },
                    false,
                  ],
                  Object {
                    "borderBottomColor": "rgba(229, 229, 229, 1)",
                    "borderLeftColor": "rgba(229, 229, 229, 1)",
                    "borderRightColor": "rgba(229, 229, 229, 1)",
                    "borderTopColor": "rgba(229, 229, 229, 1)",
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "RegularFont",
                        "fontSize": 16,
                        "fontWeight": "400",
                        "lineHeight": 24,
                      },
                      Array [
                        Object {
                          "flexBasis": "0%",
                          "flexGrow": 1,
                          "flexShrink": 1,
                          "fontSize": 14,
                          "lineHeight": 20,
                        },
                        Object {
                          "color": "rgba(82, 82, 82, 1)",
                        },
                      ],
                    ]
                  }
                >
                  foo
                </Text>
              </View>
              <View
                style={
                  Object {
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "RegularFont",
                        "fontSize": 16,
                        "fontWeight": "400",
                        "lineHeight": 24,
                      },
                      Array [
                        Object {
                          "fontFamily": "BoldFont",
                          "fontSize": 14,
                          "fontWeight": "700",
                          "lineHeight": 20,
                          "width": "8.333333%",
                        },
                        Object {
                          "color": "rgba(0, 0, 0, 1)",
                        },
                      ],
                    ]
                  }
                >
                  â€¢
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "RegularFont",
                        "fontSize": 16,
                        "fontWeight": "400",
                        "lineHeight": 24,
                      },
                      Array [
                        Object {
                          "flexBasis": "0%",
                          "flexGrow": 1,
                          "flexShrink": 1,
                          "fontSize": 14,
                          "lineHeight": 20,
                        },
                        Object {
                          "color": "rgba(82, 82, 82, 1)",
                        },
                      ],
                    ]
                  }
                >
                  bar
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
